client/src/language/middlewareReferences.ts

```diff
@@ -0,0 +1,75 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) 2023 Savoir-faire Linux. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import { type ReferencesMiddleware } from 'vscode-languageclient'
+import { type EmbeddedLanguageDocInfos, embeddedLanguageDocsManager } from './EmbeddedLanguageDocsManager'
+import { requestsManager } from './RequestManager'
+import { type Location, commands, workspace, type TextDocument } from 'vscode'
+import { getEmbeddedLanguageDocPosition, getOriginalDocRange } from './utils'
+
+export const middlewareProvideReferences: ReferencesMiddleware['provideReferences'] = async (document, position, options, token, next) => {
+  const nextResult = await next(document, position, options, token)
+  if (nextResult !== undefined && nextResult !== null) {
+    return nextResult
+  }
+  const embeddedLanguageType = await requestsManager.getEmbeddedLanguageTypeOnPosition(document.uri.toString(), position)
+  if (embeddedLanguageType === undefined || embeddedLanguageType === null) {
+    return
+  }
+  const embeddedLanguageDocInfos = embeddedLanguageDocsManager.getEmbeddedLanguageDocInfos(document.uri, embeddedLanguageType)
+  if (embeddedLanguageDocInfos === undefined || embeddedLanguageDocInfos === null) {
+    return
+  }
+  const embeddedLanguageTextDocument = await workspace.openTextDocument(embeddedLanguageDocInfos.uri)
```

Comment:
Just to make sure: is this opened/closed for references, completion, hover and definition?
Or is it just opened once, then closed once when a file analyse happens (desired).

Reply:
According to the [type's documentation]:
```
* Opens a document. Will return early if this document is already open. Otherwise
* the document is loaded and the {@link workspace.onDidOpenTextDocument didOpen}-event fires.
....
* *Note* that the lifecycle of the returned document is owned by the editor and not by the extension. That means an
* {@linkcode workspace.onDidCloseTextDocument onDidClose}-event can occur at any time after opening it.
```

So according to my understanding and experimentation, the TextDocument will open first open once when `openTextDocument` is called. Afterward, if the function is called again, the document will not be reopened, since it is already opened. The function will simply return the TextDocument.

However, the TextDocument can be closed anytime by the editor. According to my observations, it will happen after a certain amount of time after the opening, if no further action is made. According to our previous bug were the documents would reopen non stop, it seems to also happen when the user has already opened a certain amount of documents.

The middlewares (at least those we use) are called on the document on which the user is currently interacting with. However, the "embedded language documents" related to that document could have been closed, if the user had left it untouched for too long. That's the reason why openTextDocument is called no matter what.

The diagnostics can pop on documents the user is not interacting with anymore. That's why `vscode.workspace.textDocuments.find` is called instead of `openTextDocument`, to avoid reopening documents that have been closed by the editor, which previously had for result to trigger an unstoppable loop of opening and closing documents